File: tests/algorithms/x/PHP/ciphers/rail_fence_cipher.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
$__start_mem = memory_get_usage();
$__start = _now();
  function encrypt($input_string, $key) {
  global $bf;
  if ($key <= 0) {
  $panic('Height of grid can\'t be 0 or negative');
}
  if ($key == 1 || strlen($input_string) <= $key) {
  return $input_string;
}
  $lowest = $key - 1;
  $temp_grid = [];
  $i = 0;
  while ($i < $key) {
  $temp_grid = _append($temp_grid, []);
  $i = $i + 1;
};
  $position = 0;
  while ($position < strlen($input_string)) {
  $num = $position % ($lowest * 2);
  $alt = $lowest * 2 - $num;
  if ($num > $alt) {
  $num = $alt;
}
  $row = $temp_grid[$num];
  $row = _append($row, $input_string[$position]);
  $temp_grid[$num] = $row;
  $position = $position + 1;
};
  $output = '';
  $i = 0;
  while ($i < $key) {
  $row = $temp_grid[$i];
  $j = 0;
  while ($j < count($row)) {
  $output = $output . $row[$j];
  $j = $j + 1;
};
  $i = $i + 1;
};
  return $output;
};
  function decrypt($input_string, $key) {
  global $bf;
  if ($key <= 0) {
  $panic('Height of grid can\'t be 0 or negative');
}
  if ($key == 1) {
  return $input_string;
}
  $lowest = $key - 1;
  $counts = [];
  $i = 0;
  while ($i < $key) {
  $counts = _append($counts, 0);
  $i = $i + 1;
};
  $pos = 0;
  while ($pos < strlen($input_string)) {
  $num = $pos % ($lowest * 2);
  $alt = $lowest * 2 - $num;
  if ($num > $alt) {
  $num = $alt;
}
  $counts[$num] = $counts[$num] + 1;
  $pos = $pos + 1;
};
  $grid = [];
  $counter = 0;
  $i = 0;
  while ($i < $key) {
  $length = $counts[$i];
  $slice = substr($input_string, $counter, $counter + $length - $counter);
  $row = [];
  $j = 0;
  while ($j < strlen($slice)) {
  $row = _append($row, $slice[$j]);
  $j = $j + 1;
};
  $grid = _append($grid, $row);
  $counter = $counter + $length;
  $i = $i + 1;
};
  $indices = [];
  $i = 0;
  while ($i < $key) {
  $indices = _append($indices, 0);
  $i = $i + 1;
};
  $output = '';
  $pos = 0;
  while ($pos < strlen($input_string)) {
  $num = $pos % ($lowest * 2);
  $alt = $lowest * 2 - $num;
  if ($num > $alt) {
  $num = $alt;
}
  $output = $output . $grid[$num][$indices[$num]];
  $indices[$num] = $indices[$num] + 1;
  $pos = $pos + 1;
};
  return $output;
};
  function bruteforce($input_string) {
  global $bf;
  $results = [];
  $key_guess = 1;
  while ($key_guess < strlen($input_string)) {
  $results[$key_guess] = decrypt($input_string, $key_guess);
  $key_guess = $key_guess + 1;
};
  return $results;
};
  echo rtrim(encrypt('Hello World', 4)), PHP_EOL;
  echo rtrim(decrypt('HWe olordll', 4)), PHP_EOL;
  $bf = bruteforce('HWe olordll');
  echo rtrim($bf[4]), PHP_EOL;
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;

File: tests/algorithms/x/Kotlin/ciphers/rail_fence_cipher.kt
import java.math.BigInteger

fun panic(msg: String): Nothing { throw RuntimeException(msg) }

fun <T> _listSet(lst: MutableList<T>, idx: Int, v: T) { while (lst.size <= idx) lst.add(v); lst[idx] = v }

var _nowSeed = 0L
var _nowSeeded = false
fun _now(): Long {
    if (!_nowSeeded) {
        System.getenv("MOCHI_NOW_SEED")?.toLongOrNull()?.let {
            _nowSeed = it
            _nowSeeded = true
        }
    }
    return if (_nowSeeded) {
        _nowSeed = (_nowSeed * 1664525 + 1013904223) % 2147483647
        kotlin.math.abs(_nowSeed)
    } else {
        kotlin.math.abs(System.nanoTime())
    }
}

fun toJson(v: Any?): String = when (v) {
    null -> "null"
    is String -> "\"" + v.replace("\"", "\\\"") + "\""
    is Boolean, is Number -> v.toString()
    is Map<*, *> -> v.entries.joinToString(prefix = "{", postfix = "}") { toJson(it.key.toString()) + ":" + toJson(it.value) }
    is Iterable<*> -> v.joinToString(prefix = "[", postfix = "]") { toJson(it) }
    else -> toJson(v.toString())
}

fun encrypt(input_string: String, key: Int): String {
    if (key <= 0) {
        panic("Height of grid can't be 0 or negative")
    }
    if ((key == 1) || (input_string.length <= key)) {
        return input_string
    }
    var lowest: Int = key - 1
    var temp_grid: MutableList<MutableList<String>> = mutableListOf<MutableList<String>>()
    var i: Int = 0
    while (i < key) {
        temp_grid = run { val _tmp = temp_grid.toMutableList(); _tmp.add(mutableListOf<String>()); _tmp }
        i = i + 1
    }
    var position: Int = 0
    while (position < input_string.length) {
        var num: BigInteger = ((Math.floorMod(position, (lowest * 2))).toBigInteger())
        var alt = ((lowest * 2)).toBigInteger().subtract((num).toBigInteger())
        if (num.compareTo((alt as Int).toBigInteger()) > 0) {
            num = (alt as Int).toBigInteger()
        }
        var row: MutableList<String> = temp_grid[(num).toInt()]!!
        row = run { val _tmp = row.toMutableList(); _tmp.add(input_string.substring(position, position + 1)); _tmp }
        _listSet(temp_grid, (num).toInt(), (row as MutableList<String>))
        position = position + 1
    }
    var output: String = ""
    i = 0
    while (i < key) {
        var row: MutableList<String> = temp_grid[i]!!
        var j: Int = 0
        while (j < row.size) {
            output = output + row[j]!!
            j = j + 1
        }
        i = i + 1
    }
    return output
}

fun decrypt(input_string: String, key: Int): String {
    if (key <= 0) {
        panic("Height of grid can't be 0 or negative")
    }
    if (key == 1) {
        return input_string
    }
    var lowest: Int = key - 1
    var counts: MutableList<Int> = mutableListOf<Int>()
    var i: Int = 0
    while (i < key) {
        counts = run { val _tmp = counts.toMutableList(); _tmp.add(0); _tmp }
        i = i + 1
    }
    var pos: Int = 0
    while (pos < input_string.length) {
        var num: BigInteger = ((Math.floorMod(pos, (lowest * 2))).toBigInteger())
        var alt = ((lowest * 2)).toBigInteger().subtract((num).toBigInteger())
        if (num.compareTo((alt as Int).toBigInteger()) > 0) {
            num = (alt as Int).toBigInteger()
        }
        _listSet(counts, (num).toInt(), counts[(num).toInt()]!! + 1)
        pos = pos + 1
    }
    var grid: MutableList<MutableList<String>> = mutableListOf<MutableList<String>>()
    var counter: Int = 0
    i = 0
    while (i < key) {
        var length: Int = counts[i]!!
        var slice: String = input_string.substring(counter, counter + length)
        var row: MutableList<String> = mutableListOf<String>()
        var j: Int = 0
        while (j < slice.length) {
            row = run { val _tmp = row.toMutableList(); _tmp.add(slice[j].toString()); _tmp }
            j = j + 1
        }
        grid = run { val _tmp = grid.toMutableList(); _tmp.add(row); _tmp }
        counter = counter + length
        i = i + 1
    }
    var indices: MutableList<Int> = mutableListOf<Int>()
    i = 0
    while (i < key) {
        indices = run { val _tmp = indices.toMutableList(); _tmp.add(0); _tmp }
        i = i + 1
    }
    var output: String = ""
    pos = 0
    while (pos < input_string.length) {
        var num: BigInteger = ((Math.floorMod(pos, (lowest * 2))).toBigInteger())
        var alt = ((lowest * 2)).toBigInteger().subtract((num).toBigInteger())
        if (num.compareTo((alt as Int).toBigInteger()) > 0) {
            num = (alt as Int).toBigInteger()
        }
        output = output + (((grid[(num).toInt()]!!) as MutableList<String>))[indices[(num).toInt()]!!]!!
        _listSet(indices, (num).toInt(), indices[(num).toInt()]!! + 1)
        pos = pos + 1
    }
    return output
}

fun bruteforce(input_string: String): MutableMap<Int, String> {
    var results: MutableMap<Int, String> = mutableMapOf<Int, String>()
    var key_guess: Int = 1
    while (key_guess < input_string.length) {
        (results)[key_guess] = decrypt(input_string, key_guess)
        key_guess = key_guess + 1
    }
    return results
}

fun main() {
    run {
        System.gc()
        val _startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _start = _now()
        println(encrypt("Hello World", 4))
        println(decrypt("HWe olordll", 4))
        var bf: MutableMap<Int, String> = bruteforce("HWe olordll")
        println((bf)[4] as String)
        System.gc()
        val _end = _now()
        val _endMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _durationUs = (_end - _start) / 1000
        val _memDiff = kotlin.math.abs(_endMem - _startMem)
        val _res = mapOf("duration_us" to _durationUs, "memory_bytes" to _memDiff, "name" to "main")
        println(toJson(_res))
    }
}
